Sources/NIOCore/ChannelHandler.swift

```diff
@@ -354,25 +354,42 @@ public struct NIOProtocolNegotiationResult<NegotiationResult: Sendable> {
         case deferredResult(EventLoopFuture<NIOProtocolNegotiationResult<NegotiationResult>>)
     }
 
-    private let result: Result
+    private let _result: Result
 
+    /// The final result of protocol negotiation.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    public var result: NegotiationResult {
+        get async throws {
+            switch self._result {
+            case .finished(let negotiationResult):
+                return negotiationResult
+            case .deferredResult(let eventLoopFuture):
+                return try await eventLoopFuture.flatMap { $0._result.resolve(on: eventLoopFuture.eventLoop) }.get()
+            }
+        }
+    }
+
+    /// Intializes a new ``NIOProtocolNegotiationResult`` with a final result.
+    ///
+    /// - Parameter result: The final result of protocol negotiation.
     public init(result: NegotiationResult) {
-        self.result = .finished(result)
+        self._result = .finished(result)
     }
 
+    /// Intializes a new ``NIOProtocolNegotiationResult`` with a deferred result.
+    ///
+    /// - Parameter deferredResult: The deferred result.
     public init(deferredResult: EventLoopFuture<NIOProtocolNegotiationResult<NegotiationResult>>) {
-        self.result = .deferredResult(deferredResult)
+        self._result = .deferredResult(deferredResult)
     }
+}
 
-    /// Waits for the final protocol negotiation result and  returns the value.
-    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
-    public func waitForFinalResult() async throws -> NegotiationResult {
```

Comment:
Do we use this anywhere? Might be easier to deprecate and remove later to avoid breaking our other repos etc.